File: DataStructure/BinaryTree/pathSum/129_sumNumbers.cpp
// 4-23
// 跟第一次思路一样 
class Solution {
public:
    int ans = 0;
    void dfs(TreeNode* root, int sum) {
        if (root == nullptr)
            return;
        sum = sum * 10 + root->val;
        if (root->left == nullptr && root->right == nullptr) {
            ans += sum;
            return;
        }
        dfs(root->left, sum);
        dfs(root->right, sum);
    }
    int sumNumbers(TreeNode* root) {
        dfs(root, 0);
        return ans;
    }
};

File: DataStructure/BinaryTree/pathSum/129_sumNumbers.java
package LeetCode.Binary_Tree;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.ArrayList;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

// 就是深度优先遍历每一条路径
class Solution {
    int ans = 0;

    public int sumNumbers(TreeNode root) {
        dfs(root, 0);
        return ans;
    }
    
    public void dfs(TreeNode root, int sum) {
        if (root == null) {
            return;
        }
        sum = sum * 10 + root.val;
        if (root.left == null && root.right == null) {
            ans += sum;
            return;
        }

        //if (root.left != null) 这判断没用
        dfs(root.left, sum);
        //if (root.right != null)
        dfs(root.right, sum);
    }
}
// 通过了 一定要注意递归时不要用脑子压栈
// 我这是preorder postorder也是可以的